interp/numTok.ml

```diff
@@ -1,52 +1,219 @@
-    | (('e'|'E') as e) :: (('+'|'-') as sign) :: _ ->
-       begin match Stream.npeek 3 s with
-       | _ :: _ :: ('0'..'9' as c) :: _ ->
-          Stream.junk s; Stream.junk s; Stream.junk s;
-          get_buff (number (store (store (store 0 e) sign) c) s)
-       | _ -> ""
-       end
-    | _ -> "" in
-  { int = i; frac = f; exp = e }
-
-let of_string s =
-  if s = "" || s.[0] < '0' || s.[0] > '9' then None else
-    let strm = Stream.of_string (s ^ "  ") in
-    let n = parse strm in
-    if Stream.count strm >= String.length s then Some n else None
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2019       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(** We keep the string to preserve the user representation,
+    e.g. "e"/"E" or the presence of leading 0s, or the presence of a +
+    in the exponent *)
+
+module UnsignedNat =
+struct
+  type t = string
+  let of_string s =
+    assert (String.length s > 0);
+    assert (s.[0] <> '-');
+    String.(concat "" (split_on_char '_' s))
+  let to_string s = s
+
+  (** Comparing two raw numbers (base 10, big-endian, non-negative).
+      A bit nasty, but not critical: only used to decide when a
+      number is considered as large (see warnings above). *)
+
+  exception Comp of int
+
+  let rec compare s s' =
+    let l = String.length s and l' = String.length s' in
+    if l < l' then - compare s' s
+    else
+      let d = l-l' in
+      try
+        for i = 0 to d-1 do if s.[i] != '0' then raise (Comp 1) done;
+        for i = d to l-1 do
+          let c = Util.pervasives_compare s.[i] s'.[i-d] in
+          if c != 0 then raise (Comp c)
+        done;
+        0
+      with Comp c -> c
+end
+
+type sign = SPlus | SMinus
+
+module SignedNat =
+struct
+  type t = sign * UnsignedNat.t
+  let of_string n =
+    let sign,n =
+      if String.length n > 0 && n.[0] = '-'
+      then (SMinus,String.sub n 1 (String.length n - 1))
+      else (SPlus,n) in
+    (sign,UnsignedNat.of_string n)
+  let to_string (s,n) =
+    (match s with SPlus -> "" | SMinus -> "-") ^ UnsignedNat.to_string n
+  let to_bigint s = Bigint.of_string (to_string s)
+  let of_bigint n =
+    let sign, n = if Bigint.is_strictly_neg n then (SMinus, Bigint.neg n) else (SPlus, n) in
+    (sign, Bigint.to_string n)
+end
+
+module Unsigned =
+struct
+
+  type t = {
+    int : string;  (** -?\[0-9\]\[0-9_\]* *)
```

Comment:
`-?` for unsigned numbers?

Reply:
This is remaining from a previous step of the reworking of `NumTok`, I fixed.